### PR TITLE
Add cpu device parser to validate cpu devices

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Added a CPU device parser. This will help raise errors if incorrect CPU devices are selected just like tpus and gpus.
previously no errors were raised for -1, 0 and [7] and was falling back to 1 silently.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 1]
```diff
@@ -392,6 +392,7 @@ def _check_device_config_and_set_final_flags(
         if self._devices_flag in ([], 0, "0"):
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:392]: why do we fall back here? shouldn't it be strict if someone's settings `0, [], '0'` explicitly?

Comment (kaushikb11) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:392]: Nope! The logic hasn't been changed with the Accelerator Connector rewrite.

Also, a lot of users do `Trainer(accelerator="gpu", devices=torch.cuda.device_count())`, this should fall back to CPUAccelerator.

Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:392]: why not fail? it's a classic example if misconfiguration no? I am trying to use something, but if I don't have it, framework should tell me.

File: pytorch_lightning/accelerators/cpu.py [hunk 2]
```diff
@@ -39,19 +39,16 @@ def get_device_stats(self, device: _DEVICE) -> Dict[str, Any]:
         return {}
 
     @staticmethod
-    def parse_devices(devices: Union[int, str, List[int]]) -> Union[int, str, List[int]]:
+    def parse_devices(devices: Union[int, str, List[int]]) -> int:
         """Accelerator device parsing logic."""
+        devices = device_parser.parse_cpu_cores(devices)
+        assert devices is not None
```
Comment (ananthsub) [inline pytorch_lightning/accelerators/cpu.py:45]:
```suggestion
        if devices is None:
            raise MisconfigurationError("...")
```

File: pytorch_lightning/accelerators/cpu.py [hunk 3]
```diff
@@ -39,19 +39,16 @@ def get_device_stats(self, device: _DEVICE) -> Dict[str, Any]:
...
-    def get_parallel_devices(devices: Union[int, str, List[int]]) -> List[torch.device]:
+    def get_parallel_devices(devices: int) -> List[torch.device]:
         """Gets parallel devices for the Accelerator."""
-        if isinstance(devices, int):
-            return [torch.device("cpu")] * devices
-        rank_zero_warn(
-            f"The flag `devices` must be an int with `accelerator='cpu'`, got `devices={devices!r}` instead."
-        )
-        return []
+        return [torch.device("cpu")] * devices
```
Comment (ananthsub) [inline pytorch_lightning/accelerators/cpu.py:51]: this should independently validate that devices is a positive integer as well, right? what if someone doesn't call parse_devices before calling get_parallel_devices ?

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:392]: this would fail with the accelerator availability check. i agree with @rohitgr7 that we shouldn't fallback in this instance. in this case, "auto" is a better choice for users to select if they want whatever hardware is available

File: pytorch_lightning/accelerators/cpu.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/accelerators/cpu.py:45]: this was just for mypy

File: pytorch_lightning/utilities/device_parser.py [hunk 4]
```diff
@@ -137,6 +138,27 @@ def parse_tpu_cores(tpu_cores: Optional[Union[int, str, List[int]]]) -> Optional
     return tpu_cores
 
 
+def parse_cpu_cores(cpu_cores: Optional[Union[int, str, List[int]]]) -> Optional[int]:
```
Comment (ananthsub) [inline pytorch_lightning/utilities/device_parser.py:141]: why is the input typed as optional if CPUAccelerator.parse_devices doesn't accept optional inputs?

Comment (rohitgr7) [reply pytorch_lightning/utilities/device_parser.py:141]: good catch!

File: pytorch_lightning/accelerators/cpu.py (hunk 3 above)
Comment (rohitgr7) [reply pytorch_lightning/accelerators/cpu.py:51]: well, this can happen anywhere, even in `gpu.py`, `tpu.py`.. etc. I guess the assumption here was that parse_devices should be called first..
or we can parse them again inside `get_parallel_devices` for all?

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:392]: Yes, we don't need this leniency now that `auto` is supported

File: tests/accelerators/test_cpu.py [hunk 5]
```diff
@@ -66,3 +67,10 @@ def load_checkpoint(self, checkpoint_path: Union[str, Path]) -> Dict[str, Any]:
     for func in (trainer.test, trainer.validate, trainer.predict):
         plugin.setup_called = False
         func(model, ckpt_path=checkpoint_path)
+
+
+@pytest.mark.parametrize("devices", ([3], -1, "1"))
```
Comment (rohitgr7) [inline tests/accelerators/test_cpu.py:72]:
```suggestion
@pytest.mark.parametrize("devices", ([3], -1, 0))
```

File: pytorch_lightning/utilities/device_parser.py [hunk 6]
```diff
@@ -137,6 +138,30 @@ def parse_tpu_cores(tpu_cores: Optional[Union[int, str, List[int]]]) -> Optional
     return tpu_cores
 
 
+def parse_cpu_cores(cpu_cores: Union[int, str, List[int]]) -> Optional[int]:
+    """
+    Parses the cpu_cores given in the format as accepted by the
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/device_parser.py:143]:
```suggestion
    Parses the cpu_cores given in the format as accepted by `devices` flag to
```

File: pytorch_lightning/utilities/device_parser.py [hunk 7]
```diff
@@ -137,6 +138,30 @@ def parse_tpu_cores(tpu_cores: Optional[Union[int, str, List[int]]]) -> Optional
...
+
+    Returns:
+        an int representing the number of processes
+
+    Raises:
+        MisconfigurationException:
+            If cpu_cores is not an int > 0
+    """
+    if isinstance(cpu_cores, str) and cpu_cores.strip().isdigit():
+        cpu_cores = int(cpu_cores.strip())
```
Comment (carmocca) [inline pytorch_lightning/utilities/device_parser.py:157]:
```python
int("   0  ")
0
int("   0\n  ")
0
int("   1\n  ")
1
```

```suggestion
        cpu_cores = int(cpu_cores)
```

File: pytorch_lightning/utilities/device_parser.py [hunk 8]
```diff
@@ -137,6 +138,30 @@ def parse_tpu_cores(tpu_cores: Optional[Union[int, str, List[int]]]) -> Optional
     return tpu_cores
 
 
+def parse_cpu_cores(cpu_cores: Union[int, str, List[int]]) -> Optional[int]:
```
Comment (carmocca) [inline pytorch_lightning/utilities/device_parser.py:141]:
```suggestion
def parse_cpu_cores(cpu_cores: Union[int, str, List[int]]) -> int:
```

File: pytorch_lightning/accelerators/cpu.py [hunk 9]
```diff
@@ -39,19 +39,17 @@ def get_device_stats(self, device: _DEVICE) -> Dict[str, Any]:
         return {}
 
     @staticmethod
-    def parse_devices(devices: Union[int, str, List[int]]) -> Union[int, str, List[int]]:
+    def parse_devices(devices: Union[int, str, List[int]]) -> int:
         """Accelerator device parsing logic."""
+        devices = device_parser.parse_cpu_cores(devices)
+        assert devices is not None
```
Comment (carmocca) [inline pytorch_lightning/accelerators/cpu.py:45]: If `Optional[int]` is fixed from the `parse_cpu_cores` then we wouldn't need this assertion

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/utilities/device_parser.py (hunk 7 above)
Comment (rohitgr7) [reply pytorch_lightning/utilities/device_parser.py:157]: TIL

**Review by justusschock — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

File: pytorch_lightning/utilities/parsing.py [hunk 10]
```diff
@@ -243,7 +243,7 @@ def save_hyperparameters(
         if isinstance(init_args[k], nn.Module):
             rank_zero_warn(
                 f"Attribute {k!r} is an instance of `nn.Module` and is already saved during checkpointing."
-                " It is recommended to ignore them using `self.save_hyperparameters(ignore=[k!r])`."
+                f" It is recommended to ignore them using `self.save_hyperparameters(ignore=[{k}!r])`."
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply pytorch_lightning/utilities/parsing.py:246]: you are 100% correct!
